gnomad/utils/constraint.py

```diff
@@ -360,3 +412,274 @@ def collapse_strand(
+        log_coverage_expr=low_cov_group_ht.log_coverage,
+    )
+    coverage_model = tuple(low_cov_group_ht.aggregate(coverage_model_expr).beta)
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
+    cpg_expr: hl.expr.BooleanExpression,
+    mu_snp_expr: hl.expr.Float64Expression,
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_exprs: Dict[str, hl.ArrayNumericExpression] = {},
+    weighted: bool = False,
+) -> Dict[str, Union[Dict[bool, hl.expr.ArrayExpression], hl.ArrayExpression]]:
+    """
+    Build plateau models to calibrate mutation rate to compute predicted proportion observed value.
+
+    The x and y of the plateau models:
+    - x: `mu_snp` - mutation rate
+    - y: proportion observed ('observed_variants' or 'observed_{pop}' / 'possible_variants')
+
+    :param cpg_expr: BooleanExpression noting whether a site is a CPG site.
+    :param mu_snp_expr: Float64Expression of the mutation rate.
+    :param observed_variants_expr: Int64Expression of the observed variant counts
+      for each combination of keys in `ht`.
+    :param possible_variants_expr: Int64Expression of the possible variant counts
+      for each combination of keys in `ht`.
+    :param pop_observed_variants_exprs: Dictionary with population names (keys)
+      and observed variant counts ArrayNumericExpressions (values) for
+      specified populations. Default is {}.
+    :param weighted: Whether to generalize the model to weighted least squares using
+      'possible_variants'. Default is False.
+    :return: A dictionary of intercepts and slopes for plateau models of each
+      population. The key of the dictionary is a population name, and the value is a
+      dictionary (or a list of dictionaries if `pop_observed_variants_exprs` is
+      specified) mapping cpg BooleanExpression to a intercept and a slope.
+    """
+    # Build a plateau model using all the sites in the Table.
+    plateau_models_agg_expr = {
+        "total": hl.agg.group_by(
+            cpg_expr,
+            hl.agg.linreg(
+                observed_variants_expr / possible_variants_expr,
+                [1, mu_snp_expr],
+                weight=possible_variants_expr if weighted else None,
+            ).beta,
+        )
+    }
+    # Build plateau models using sites in population downsamplings if
+    # population is specified.
+    plateau_models_agg_expr = {
+        pop: hl.agg.array_agg(
+            lambda pop_observed_variants: hl.agg.group_by(
+                cpg_expr,
+                hl.agg.linreg(
+                    pop_observed_variants / possible_variants_expr,
+                    [1, mu_snp_expr],
+                    weight=possible_variants_expr if weighted else None,
+                ).beta,
+            ),
+            pop_observed_variants_expr,
+        )
+        for pop, pop_observed_variants_expr in pop_observed_variants_exprs.items()
```

Comment:
I ended up using dictionary comprehension. There is a hail discussion [here](https://hail.zulipchat.com/#narrow/stream/123010-Hail-Query-0.2E2-support/topic/Iterate.20through.20DictExpression).